projects/laji/src/app/shared-modules/observation-map/observation-map/observation-map.component.ts

```diff
@@ -209,210 +237,272 @@ export class ObservationMapComponent implements OnChanges, OnDestroy {
-    const zoomedOut = this.activeZoomThresholdLevel < curActiveZoomThresholdLevel;
-
-    if (alreadyShowingAll) {
-      return;
-    }
+    const activeBoundsContainViewport = this.activeZoomThresholdBounds && this.activeZoomThresholdBounds.contains(e.bounds);
 
-    if (outOfActiveBounds
-      || zoomedInAndNotShowingPoints
-      || zoomedOut
+    if (
+      !(this.showingIndividualPoints && activeBoundsContainViewport)
+      && (
+           // Zoom in
+           this.activeZoomThresholdLevel > curActiveZoomThresholdLevel
+           // Zoom out
+        || this.activeZoomThresholdLevel < curActiveZoomThresholdLevel
+           // Move
+        || !activeBoundsContainViewport
+      )
     ) {
       this.activeZoomThresholdBounds = e.bounds.pad(1);
-      this.updateMapData();
+      this.updateMap();
     }
   }
 
-  initLegendTopMargin(): void {
-    const top = 20, items = this.color instanceof Array ? this.color.length : 1;
-    this.topMargin = '-' + (top + (items * 20)) + 'px';
+  onTileLayersChange(layerOptions: TileLayersOptions) {
+    const temp = layerOptions.active === 'finnish';
+    if (temp !== this.useFinnishMap) {
+      this.useFinnishMap = temp;
+      this.updateMap();
+    }
   }
 
-  initLegend(): void {
-    const legend = [];
-    let start = 1;
-    if (this.color instanceof Array) {
-      this.color.map((color, idx) => {
-        let end = '+', newStart;
-        if (this.colorThresholds[idx]) {
-          newStart = this.colorThresholds[idx];
-          end = '-' + newStart;
-        }
-        legend.push({
-          color,
-          range: start + end
-        });
-        start = newStart + 1;
-      });
+  onVisualizationModeChange(mode: string) {
+    const fetchRequired = mode !== this.visualizationMode && (
+      mode === 'redlistStatus'
+      || this.visualizationMode === 'redlistStatus'
+    );
+    this.visualizationMode = <ObservationVisualizationMode>mode;
+
+    if (fetchRequired) {
+      this.updateMap();
+    } else {
+      const dataOptions = this.mapData[0];
+      if (!dataOptions) { return; }
+      const vis = this.visualization?.[this.visualizationMode];
+      if (vis?.getFeatureStyle) { dataOptions.getFeatureStyle = vis.getFeatureStyle; }
+      // changing mapData object reference so that laji-map recognizes that it has changed
+      this.mapData = [...this.mapData];
     }
-    this.legendList = legend;
-    this.changeDetector.markForCheck();
   }
 
-  refreshMap() {
-    setTimeout(() => {
-      this.reloading = true;
-    }, 100);
+  resetTable() {
+    this.selectedObservationCoordinates = undefined;
+    this.cdr.detectChanges();
+    // Wait until next cycle so that height has time to adjust to the removal of the table
     setTimeout(() => {
-      this.reloading = false;
-    }, 200);
+      this.tableViewHeightOverride = undefined;
+      this.cdr.detectChanges();
+    });
```

Comment:
Did you try could this be done at `AfterViewChecked` so no need for timeout?

Reply:
Idk it might work, but I don't see why this would be an improvement? Calling detectChanges upon every view check is unnecessarily bad for performance, or alternatively there could be a flag to tell if it should be called on the next viewCheck, but that would complicate the code unnecessarily. An empty `setTimeout` isn't evil really, it just creates a callback in the js event loop. To be fair it's not at all obvious why it's required, but I think the alternatives are worse

Reply:
It's not that bad as it is, we can go with this. I agree it would become more hard to follow with a flag. With the comment it's understandable.

In general I'm always worried about setTimeouts maintenance wise for many reasons. It's just a single tiny setTimeout which can seem innocent, but I don't like that it assumes that the table height is updated synchronously - if that changes this will break and it will be challenging to debug. It's difficult to find inline asynchronous operations in code. The `AfterViewChecked` should be guaranteened to run after the the actually is updated, even if it becomes async. Also, if there appear more setTimeouts, the order of execution can be really chaotic to follow.